cassiopeia/data/CassiopeiaTree.py

```diff
@@ -1334,14 +1651,130 @@ def filter_nodes(self, condition: Callable[[str], bool]) -> List[str]:
+                "At least two distinct nodes must be provided"
+            )
+
+        # Reversing get_all_ancestors gives a list of nodes from the root to each
+        # node. Since get_all_ancestors doesn't include the node itself, we add
+        # those manually.
+        all_ancestors = [
+            reversed(self.get_all_ancestors(node, include_node=True))
+            for node in nodes
+        ]
+        last_ancestor = self.root
+        for ancestors in zip(*all_ancestors):
+            if len(set(ancestors)) > 1:
+                break
+            last_ancestor = ancestors[0]
+        return last_ancestor
+
+    def get_distance(self, node1: str, node2: str) -> float:
+        """Compute the branch distance between two nodes.
+
+        Internally, this function converts the tree to an undirected graph and
+        finds the shortest path by calling :func:`nx.shortest_path_length`.
+
+        Args:
+            node1: First node
+            node2: Second node
+
+        Returns:
+            The branch distance between the two nodes
+
+        Raises:
+            CassiopeiaTreeError if the tree has not been initialized.
+        """
+        self.__check_network_initialized()
+
+        key = frozenset({node1, node2})
+        if (
+            "distances" not in self.__cache
+            or key not in self.__cache["distances"]
+        ):
+            distances = self.__cache.setdefault("distances", {})
+            undirected_network = self.__network.to_undirected()
+            distances[key] = nx.shortest_path_length(
+                undirected_network, source=node1, target=node2, weight="time"
+            )
+
+        return self.__cache["distances"][key]
+
+    def get_distances(
+        self, node: str, leaves_only: bool = False
+    ) -> Dict[str, float]:
+        """Compute the distances between the given node and every other node.
+
+        Internally, this function converts the tree to an undirected graph and
+        finds the shortest path by calling :func:`nx.shortest_path_length`.
+
+        Args:
+            node: Node from which to compute distance to all other nodes
+            leaves_only: Calculate distances to leaves only and not internal nodes.
+                Defaults to False.
+
+        Returns:
+            Dictionary of distances
+
+        Raises:
+            CassiopeiaTreeError if the tree has not been initialized.
+        """
+        self.__check_network_initialized()
+
+        if "distances" not in self.__cache or any(
+            frozenset({node, _node}) not in self.__cache["distances"]
+            for _node in self.nodes
+        ):
+            distances = self.__cache.setdefault("distances", {})
+            undirected_network = self.__network.to_undirected()
+            for other_node, distance in nx.shortest_path_length(
+                undirected_network, source=node, weight="time"
+            ).items():
+                distances[frozenset({node, other_node})] = distance
+
```

Comment:
Should the weight not be "length" if we are to be weighting over branch lengths? "time" refers to the node time.